File: vol3/vol3-java-examples/src/main/java/com/heatonresearch/aifh/examples/ann/LearnXORRPROP.java
/*
 * Artificial Intelligence for Humans
 * Volume 3: Deep Learning and Neural Networks
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh
 *
 * Copyright 2014-2015 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */
package com.heatonresearch.aifh.examples.ann;

import com.heatonresearch.aifh.ann.BasicLayer;
import com.heatonresearch.aifh.ann.BasicNetwork;
import com.heatonresearch.aifh.ann.activation.ActivationSigmoid;
import com.heatonresearch.aifh.ann.train.ResilientPropagation;
import com.heatonresearch.aifh.general.data.BasicData;

import java.util.Arrays;
import java.util.List;


public class LearnXORRPROP {
    /**
     * The input necessary for XOR.
     */
    public static double XOR_INPUT[][] = { { 0.0, 0.0 }, { 1.0, 0.0 },
            { 0.0, 1.0 }, { 1.0, 1.0 } };

    /**
     * The ideal data necessary for XOR.
     */
    public static double XOR_IDEAL[][] = { { 0.0 }, { 1.0 }, { 1.0 }, { 0.0 } };

    /**
     * The main method.
     * @param args No arguments are used.
     */
    public static void main(final String args[]) {

        BasicNetwork network = new BasicNetwork();
        network.addLayer(new BasicLayer(null,true,2));
        network.addLayer(new BasicLayer(new ActivationSigmoid(),true,5));
        network.addLayer(new BasicLayer(new ActivationSigmoid(),false,1));
        network.finalizeStructure();
        network.reset();

        List<BasicData> trainingData = BasicData.combineXY(XOR_INPUT, XOR_IDEAL);

        // train the neural network
        final ResilientPropagation train = new ResilientPropagation(network, trainingData);

        int epoch = 1;

        do {
            train.iteration();
            System.out.println("Epoch #" + epoch + " Error:" + train.getLastError());
            epoch++;
        } while(train.getLastError() > 0.01);

        // test the neural network
        System.out.println("Neural Network Results:");
        for(int i=0;i < XOR_INPUT.length; i++ ) {
            double[] output = network.computeRegression(XOR_INPUT[i]);
            System.out.println(Arrays.toString(XOR_INPUT[i])
                    + ", actual=" + Arrays.toString(output)
                    + ",ideal=" + Arrays.toString(XOR_IDEAL[i]));
        }
    }
}

File: vol3/vol3-csharp-examples/AIFH-Vol3/Examples/ANN/LearnXORRPROP.cs
using AIFH_Vol3.Core.General.Data;
using AIFH_Vol3.Examples.Learning;
using AIFH_Vol3_Core.Core.ANN;
using AIFH_Vol3_Core.Core.ANN.Activation;
using AIFH_Vol3_Core.Core.ANN.Train;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AIFH_Vol3.Examples.ANN
{
    public class LearnXORRPROP : SimpleLearn
    {
        /// <summary>
        /// The input necessary for XOR.
        /// </summary>
        public static double[][] XOR_INPUT =
        {
            new double[] { 0.0, 0.0 },
            new double[] { 1.0, 0.0 },
            new double[] { 0.0, 1.0 },
            new double[] { 1.0, 1.0 } };

        /// <summary>
        /// The ideal data necessary for XOR.
        /// </summary>
        public static double[][] XOR_IDEAL =
        {
            new double[] { 0.0 },
            new double[] { 1.0 },
            new double[] { 1.0 },
            new double[] { 0.0 } };

        /// <summary>
        /// The name of this example.
        /// </summary>
        public static string ExampleName = "XOR with Resilient Propagation (RPROP).";

        /// <summary>
        /// The chapter this example is from.
        /// </summary>
        public static int ExampleChapter = 7;

        /// <summary>
        ///     The entry point for this example.  If you would like to make this example
        ///     stand alone, then add to its own project and rename to Main.
        /// </summary>
        /// <param name="args">Not used.</param>
        public static void ExampleMain(string[] args)
        {
            BasicNetwork network = new BasicNetwork();
            network.AddLayer(new BasicLayer(null, true, 2));
            network.AddLayer(new BasicLayer(new ActivationSigmoid(), true, 5));
            network.AddLayer(new BasicLayer(new ActivationSigmoid(), false, 1));
            network.FinalizeStructure();
            network.Reset();

            IList<BasicData> trainingData = BasicData.ConvertArrays(XOR_INPUT, XOR_IDEAL);

            // train the neural network
            ResilientPropagation train = new ResilientPropagation(network, trainingData);

            int epoch = 1;

            do
            {
                train.Iteration();
                Console.WriteLine("Epoch #" + epoch + " Error:" + train.LastError);
                epoch++;
            } while (train.LastError > 0.01);

            // test the neural network
            Console.WriteLine("Neural Network Results:");
            for (int i = 0; i < XOR_INPUT.Length; i++)
            {
                double[] output = network.ComputeRegression(XOR_INPUT[i]);
                Console.WriteLine(string.Join(",", XOR_INPUT[i])
                        + ", actual=" + string.Join(",", output)
                        + ",ideal=" + string.Join(",", XOR_IDEAL[i]));
            }
        }
    }
}
